---
title: "Ssd1 enrichment from normalizing to RNA-seq on full-length transcripts"
author: "Edward Wallace"
date: "6th Feb 2020"
output: 
  html_document:
    toc: true
---

# Summary

This script is to make figures of relative *enrichment* of Ssd1 on mRNA, by processing time & condition matched RNA seq and Ssd1 CRAC data, crudely normalized to transcripts per million (TPM) on full-length transcripts.

This script relies on CRAC transcript counts produced by `multiBamCov` running in the CRAC pipeline, and found in `Ssd1_CRAC_demult_dedup_20190114_all/multicov_analyses/allsample_transcriptcounts.txt`. So it must be run after  `src/CRAC_pipeline_SE_demult_dedup.py`, see the repository `README.md` for details. 

This script also relies on RNA-seq transcript counts in `results/RNAseq_stressmatched_transcriptcounts.txt` produced by `multiBamCov`, running in `process_RNAseq.Rmd`. So this script must be run after `process_RNAseq.Rmd`.

Some of the figures rely on the lists of enriched genes from DESeq2 analysis, produced by the script `deseq2_Ssd1_CRAC_vs_RNAseq.Rmd`:
* `Ssd1_enrichment_DeSeq2_30C_2x_padj0p05_genesonly.txt`
* `Ssd1_enrichment_DeSeq2_42C_2x_padj0p05_genesonly.txt`

This analysis shows that:

* RNA-seq reads are highly reproducible between replicates
* Ssd1 reads are quite reproducible, but have low coverage
* The highest-enriched targets of Ssd1 are mRNAs encoding cell wall proteins.
* Stress protein mRNAs that are most induced by heat shock are *not enriched* in Ssd1 binding, although they are bound by Ssd1.

Both RNAseq and Ssd1-CRAC are quantified on *full-length estimated unspliced transcripts* using bedtools multicov after genomic alignment with novoalign. This has some problems. For example, reads on 2 overlapping transcripts are counted on *both*, if the UTRs or ORFs overlap.

Note that this analysis:

* is restricted to mRNAs encoding verified ORFs froms nuclear chromosomes: no mitochondrial chromosome or ncRNAs.
* uses Pelechano 2013 "most abundant" annotation of transcript ends, and adds a default 25nt 5'UTR and 125nt 3'UTR for transcripts not annotated in Pelechano 2013.
* normalises count density by transcript length, to give Transcripts Per Million (TPM).
* has different goals and a cruder statistical model than the DESeq2 analysis in `deseq2_Ssd1_CRAC_vs_RNAseq.Rmd`.


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache=TRUE, warning=FALSE, message=FALSE,
                      cache.path = "cache/normalise_Ssd1_CRAC_counts-",
                      fig.path = "figure/normalise_Ssd1_CRAC_counts-")

library(tidyverse)
library(GGally)
library(cowplot)
theme_set(theme_cowplot(font_size=11) %+replace% 
            theme(strip.background=element_blank(),
                  panel.border = element_rect(color = "grey85", fill = NA)))

# setwd("/homes/ewallac2/Repos/Ssd1_CRACanalysis_2020/rmarkdown")

calc_cpm <- function(x) x / sum(x) * 1e6

calc_tpm <- function(ct,eff_length) {
  # calculate transcripts per million from count and effective length
  dens <- ct / eff_length
  dens_total <- sum(dens, na.rm=TRUE) 
  return( dens / dens_total * 1e6 )
}

log10plusepsilon <- function(x,epsilon=1) log10( x + epsilon )

gene_from_attr <- function(attr_string,
                           before="ID=",gene_str="[\\w-]+",after="_") {
    # pull a gene name from an attribute string
    match_str = paste0(before,gene_str,after)
    attr_string %>%
        stringr::str_extract(match_str) %>%
        stringr::str_remove(before) %>%
        stringr::str_remove(after)
}

scientific_10 <- function(x) {
    xout <- gsub("1e", "10^{", format(x),fixed=TRUE)
    xout <- gsub("{-0", "{-", xout,fixed=TRUE)
    xout <- gsub("{+", "{", xout,fixed=TRUE)
    xout <- gsub("{0", "{", xout,fixed=TRUE)
    xout <- paste(xout,"}",sep="")
    return(parse(text=xout))
}

scale_x_log10_sci <- function(name=waiver(),omag=seq(-10,20),...) {
    breaks10 <- 10^omag
    scale_x_log10(name,breaks=breaks10,labels=scientific_10(breaks10),...)
}

scale_y_log10_sci <- function(name=waiver(),omag=seq(-10,20),...) {
    breaks10 <- 10^omag
    scale_y_log10(name,breaks=breaks10,labels=scientific_10(breaks10),...)
}

scale_loglog <- function(...) {
    list(scale_x_log10_sci(...),scale_y_log10_sci(...))
}

ggpairs_genewise <- function(...) {
  ggpairs( lower = list(continuous = wrap("points", size=0.2,colour="grey50") ),
           upper = list(continuous = wrap("cor",stars = FALSE), 
                        combo = "box_no_facet", discrete = "count", 
                        na ="na"),
           ...)
}
```

# Inspecting RNA-seq count data

## Load RNA-seq count data

```{r load_RNAseq_count}
counts_RNA_stranded <- read_tsv("../results/RNAseq_stressmatched_transcriptcounts.txt",
                                comment = "#")  %>%
  mutate(Gene=gene_from_attr(attributes))

tpms_RNA_stranded <-  counts_RNA_stranded %>% 
  select(Gene, RNA_A_30,RNA_B_30,RNA_C_GE,RNA_D_GE,RNA_E_42,RNA_F_42) %>%
  mutate_if(is.numeric,
            calc_tpm,
            eff_length = counts_RNA_stranded$end - counts_RNA_stranded$start)

```

## Scatter plot shows genewise CPMS are VERY reproducible

```{r scatter_RNAseq_cpm,dependson="load_RNAseq_count",fig.height=7,fig.width=7}
counts_RNA_stranded %>%
  select(RNA_A_30,RNA_B_30,RNA_C_GE,RNA_D_GE,RNA_E_42,RNA_F_42) %>%
  mutate_all(calc_cpm) %>%
  mutate_all(log10plusepsilon) %>%
  ggpairs_genewise()
```

## Scatter plot shows genewise TPMs are likewise very reproducible

```{r scatter_RNAseq_tpm_all,dependson="load_RNAseq_count",fig.height=7,fig.width=7}
tpms_RNA_stranded %>%
  select(RNA_A_30,RNA_B_30,RNA_C_GE,RNA_D_GE,RNA_E_42,RNA_F_42) %>%
  mutate_all(log10plusepsilon,epsilon=1) %>%
  ggpairs_genewise()
```

```{r scatter_RNAseq_tpm,dependson="load_RNAseq_count",fig.height=4.5,fig.width=4.5}
tpms_RNA_stranded %>%
  select("RNA rep A\n30°C" = RNA_A_30,
         "RNA rep B\n30°C" = RNA_B_30,
         "RNA rep A\n42°C" = RNA_E_42,
         "RNA rep B\n42°C" = RNA_F_42) %>%
  mutate_all(log10plusepsilon,epsilon=1) %>%
  ggpairs_genewise()
```


# Inspecting Ssd1-bound data at genewise level

## Load Ssd1-bound genewise counts

```{r load_Ssd1_counts}
counts_Ssd1_stranded <- read_tsv("../Ssd1_CRAC_demult_dedup_20190114_all/multicov_analyses/allsample_transcriptcounts.txt",
                                 comment = "#") %>% 
  mutate(Gene=gene_from_attr(attributes))%>%
  select(Gene,seqname,feature,start,end,attributes,
         Ssd1_3_30 = `20190114_Ssd1_CRAC_trimmed_NNNGTGAGC_SSD1_3_30.bam`,
         Ssd1_4_30 = `20190114_Ssd1_CRAC_trimmed_NNNTGGAGC_SSD1_4_30.bam`,
         Ssd1_3_42 = `20190114_Ssd1_CRAC_trimmed_NNNAGAGC_SSD1_3_42.bam`,
         Ssd1_4_42 = `20190114_Ssd1_CRAC_trimmed_NNNCTAGC_SSD1_4_42.bam`, 
         BY4741 = `20190114_Ssd1_CRAC_trimmed_NNNGACTTAGC_BY4741.bam`) 

# rename: ("Ssd1_3_30","Ssd1_4_30","Ssd1_3_42","Ssd1_4_42", "BY4741"))

tpms_Ssd1_stranded <-  counts_Ssd1_stranded %>% 
  select(Gene, Ssd1_3_30,Ssd1_4_30,Ssd1_3_42,Ssd1_4_42,BY4741) %>%
  mutate_if(is.numeric,
            calc_tpm,
            eff_length = counts_Ssd1_stranded$end - counts_Ssd1_stranded$start)

```

```{r show_Ssd1_counts,dependson="load_Ssd1_counts"}
counts_Ssd1_stranded
```


## Ssd1-bound genewise CPMs are reproducible

```{r scatter_Ssd1_cpm,dependson="load_Ssd1_count",fig.height=5,fig.width=5}
counts_Ssd1_stranded %>%
  select(Ssd1_3_30,Ssd1_4_30,Ssd1_3_42,Ssd1_4_42,BY4741) %>%
  mutate_all(calc_cpm) %>%
  mutate_all(log10plusepsilon,epsilon=100) %>%
  ggpairs_genewise()
```

This is not as reproducible as the RNA-seq, due to relatively low coverage of deduplicated reads.

## Ssd1-bound genewise TPMs are reproducible

Here transform TPMs by `log10(tpm + 10)`, where the `+ 10` smooths away some of the low-count noise.

```{r scatter_Ssd1_tpm,dependson="load_Ssd1_count",fig.height=5.2,fig.width=5.2}
tpms_Ssd1_stranded %>%
  select("Ssd1-HTP (3)\n30°C" = Ssd1_3_30,
         "Ssd1-HTP (4)\n30°C" = Ssd1_4_30,
         "Ssd1-HTP (3)\n42°C" = Ssd1_3_42,
         "Ssd1-HTP (4)\n42°C" = Ssd1_4_42,
         "WT (BY4741)\n30°C" = BY4741) %>%
  mutate_all(log10plusepsilon,epsilon=10) %>%
  ggpairs_genewise()
```

# Compare Ssd1 CRAC to RNA-seq data to estimate crude enrichment

## Construct a single table containing Ssd1-bound and RNA-seq counts

We:

* melt/pivot the data to get in tidy long format
* make average TPMs per condition
* join these "average TPMs per condition" for Ssd1 and RNA
* calculate a ratio of TPMs; this is crude as it's normalised by total counts only
* add gene names from SGD
* confirm that TPM normalization correct by checking that TPMs sum to 1000000.


### Melt the data frames and summarize (assigned) TPMs per sample

```{r melt_tpm_both,dependson=c("load_RNAseq_count","load_Ssd1_count")}
tpms_RNA_melt <- 
  tpms_RNA_stranded %>%
  gather(key="Sample",value="TPM",-Gene) 

tpms_RNA_melt %>% group_by(Sample) %>% summarize(TPM=sum(TPM))

tpms_Ssd1_melt <- 
  tpms_Ssd1_stranded %>%
  gather(key="Sample",value="TPM",-Gene)

tpms_Ssd1_melt  %>% group_by(Sample) %>% summarize(TPM=sum(TPM))
```

### Calculate average TPMs and join

```{r mean_join_tpm,dependson=c("melt_tpm_both")}
scer_mrna_prot <- read_tsv("../input_annotation/scer-mrna-protein-absolute-estimate.txt",
                           comment="#") %>%
  rename(Gene=orf,Name=gene)

tpm_RNA_bycondition <- 
  tpms_RNA_melt %>%
  separate(Sample,c("SType","Rep","Condition")) %>%
  group_by(Gene,Condition) %>%
  summarize(TPM.mean = mean(TPM,na.rm=TRUE))

tpm_Ssd1_bycondition <- 
  tpms_Ssd1_melt %>%
  separate(Sample,c("SType","Rep","Condition")) %>%
  group_by(Gene,Condition) %>%
  summarize(TPM.mean = mean(TPM,na.rm=TRUE))

tpm_join_bycondition <-
  full_join( tpm_RNA_bycondition %>% filter(Condition %in% c("30","42")),
              tpm_Ssd1_bycondition %>% filter(Condition %in% c("30","42")),
              by=c("Gene","Condition"),
              suffix=c(".RNA",".Ssd1")) %>%
  mutate(ratio.Ssd1.RNA = TPM.mean.Ssd1/TPM.mean.RNA) %>%
  left_join(scer_mrna_prot %>% select(Gene,Name)) %>%
  mutate(Condition = factor(Condition,
                            levels = c("30","42"),
                            labels = c("30°C","42°C"))
  ) %>%
  select(Gene,Name,Condition,
         TPM.RNA=TPM.mean.RNA,TPM.Ssd1=TPM.mean.Ssd1,
         ratio.Ssd1.RNA)
```

### What the TPM joined data look like

```{r show_tpm_join_bycondition,dependson="mean_join_tpm"}
tpm_join_bycondition
```


## Plot TPMs by condition, Ssd1-bound vs RNA, with some labels

```{r plot_Ssd1_vs_RNA_labels,dependson="mean_join_tpm",fig.height=3.2,fig.width=6}
ggplot(data=tpm_join_bycondition,aes(x=TPM.RNA,y=TPM.Ssd1)) +
  geom_point(size=0.2,colour="grey50") +
  ggrepel::geom_text_repel(
    data=
      tpm_join_bycondition %>%
      filter(TPM.RNA > 10000 | TPM.Ssd1 > 10000 | 
               (ratio.Ssd1.RNA > 50 & TPM.Ssd1 > 500) ),
    aes(label=Name),size=2,
    label.padding=0.02,segment.size=0.1,force=0.1,
    # set random number seed so the labels go in the same place every time
    seed = 42) + 
  scale_x_log10_sci("RNA-seq density (TPM)",
                limits=c(5,max(tpm_join_bycondition$TPM.RNA)),
                oob=scales::squish,expand=c(0.01,0.01)) +
  scale_y_log10_sci("Ssd1 CRAC density (TPM)",
                limits=c(5,max(tpm_join_bycondition$TPM.Ssd1)),
                oob=scales::squish,expand=c(0.01,0.01)) +
  facet_wrap(~Condition)
```

## Ssd1-enriched genes by crude TPM ratios

These are familiar from other datasets: TOS1, TOS6, SUN4/SIM1/UTH1/SRL1, etc.

```{r enrich_Ssd1_vs_RNA,dependson="mean_join_tpm",results.show=TRUE}
tpm_join_bycondition %>%
  filter(Condition=="30°C",TPM.RNA > 20) %>%
  arrange(desc(ratio.Ssd1.RNA))

tpm_join_bycondition %>%
  filter(Condition=="42°C",TPM.RNA > 20) %>%
  arrange(desc(ratio.Ssd1.RNA))
```


### Make output files for crude Ssd1-enriched genes over 4x

We conservatively report transcripts that are at least 4-fold enriched in Ssd1 CRAC reads compared to RNA-seq, and with at least 20 TPM in the RNA-seq data. This filters out low-expressed genes with noisy counts.

```{r enrich_Ssd1_vs_RNA_out,dependson="mean_join_tpm",results.show=TRUE}
Ssd1_targets_TPMratio4x_30C <- 
  tpm_join_bycondition %>%
  filter(Condition=="30°C",TPM.RNA > 20, ratio.Ssd1.RNA > 4) %>%
  arrange(desc(ratio.Ssd1.RNA)) 

write_tsv(x = Ssd1_targets_TPMratio4x_30C %>%
            select(-Condition) %>%
            mutate_if(is.numeric, round, digits = 3),
          path="../results/Ssd1_targets_TPMratio4x_30C.txt")

Ssd1_targets_TPMratio4x_42C <- 
  tpm_join_bycondition %>%
  filter(Condition=="42°C",TPM.RNA > 20, ratio.Ssd1.RNA > 4) %>%
  arrange(desc(ratio.Ssd1.RNA))

write_tsv(x = Ssd1_targets_TPMratio4x_42C %>%
            select(-Condition) %>%
            mutate_if(is.numeric, round, digits = 3),
          path="../results/Ssd1_targets_TPMratio4x_42C.txt")
```

## Which genes are enriched at 42C not 30C?

```{r enrich_Ssd1_42more,dependson="enrich_Ssd1_vs_RNA_out"}
Ssd1_targets_TPMratio4x_42C %>% 
  filter(! Gene %in% Ssd1_targets_TPMratio4x_30C$Gene)
```

## Which genes are enriched at 30C not 42C?

```{r enrich_Ssd1_30more,dependson="enrich_Ssd1_vs_RNA_out"}
Ssd1_targets_TPMratio4x_30C %>% 
  filter(! Gene %in% Ssd1_targets_TPMratio4x_42C$Gene)
```


## The highest-expressed genes at 42C are not enriched in Ssd1 binding

```{r highest_42C,dependson="enrich_Ssd1_vs_RNA_out"}
tpm_join_bycondition %>%
  filter(Condition=="42°C") %>%
  arrange(desc(TPM.RNA))
```

These,e.g. Hsp104, aren't enriched in Ssd1 binding. 

## Highly-expressed genes at 42C *over 2x* enriched in Ssd1 binding

```{r highest_42C_ssd1_2x,dependson="mean_join_tpm"}
tpm_join_bycondition %>%
  filter(Condition=="42°C",ratio.Ssd1.RNA > 2) %>%
  arrange(desc(TPM.RNA))
```

* SPO24 again
* TSA1/SOD1 involved in oxidative stress resistance
* CCW12/CWP1/TIP1/BGL2 stress-induced cell wall proteins
* YDR524C-B Putative protein of unknown function; SWAT-GFP and mCherry fusion proteins localize to the endoplasmic reticulum


## Plot TPMs by condition, DEseq2-enriched genes labeled

This is figure panel 4C of the manuscript, output to the file `../figure_out/Ssd1CRAC_vs_RNAseq_TPM_colourlabels.pdf`.

```{r plot_Ssd1_vs_RNA_colourlabels,dependson="mean_join_tpm",fig.height=3.2,fig.width=6}
# We had a longer and harder to follow list of 16 targets, that is commented out here; slimmed down to 8 select for a more readable figure
# Ssd1_common_targets_df <-
#   readr::read_tsv("../results/Ssd1_targets_common_Hogan_Jansen_Bayne30C.txt",
#                   comment = "#")
# 
# Ssd1_targets_16select <- c(Ssd1_common_targets_df$Name, c("CCW12","CWP2","SPO24","NCA3"))

Ssd1_targets_8select <- c("CTS1","DSE2","SIM1","SUN4","UTH1","SCW4","CCW12","SRL1")

Ssd1_targets_DESeq2_2x_p0p05 <- bind_rows(
  tibble(Gene = readr::read_lines("../results/Ssd1_enrichment_DeSeq2_30C_2x_padj0p05_genesonly.txt"),
         Condition = "30"),
  tibble(Gene = readr::read_lines("../results/Ssd1_enrichment_DeSeq2_42C_2x_padj0p05_genesonly.txt"),
         Condition = "42")) %>%
  dplyr::mutate(Condition = factor(Condition,
                            levels = c("30","42"),
                            labels = c("30°C","42°C")),
                GType = "Ssd1_target"
                )

HSPs_select <- c("HSP42","HSP104","SSA1","SSA2","SSA3","SSA4")

Abundant_genes_select <- c("TDH3")

select_genetype_df <- bind_rows(
  tibble(GType = "Ssd1_target", Name = Ssd1_targets_8select),
  tibble(GType = "HSPs_select", Name = HSPs_select),
  tibble(GType = "Abundant", Name = Abundant_genes_select)
)

scale_colour_gtype <-
  scale_colour_manual(values = c(Ssd1_target = "blue2",
                                 HSPs_select = "red3",
                                 Abundant = "black") 
                      )

plot_Ssd1_vs_RNA_colourlabels <- 
ggplot(data=tpm_join_bycondition,aes(x=TPM.RNA,y=TPM.Ssd1)) +
  geom_point(size=0.15,colour="grey50") +
  geom_point(
    data = left_join(Ssd1_targets_DESeq2_2x_p0p05,
                     tpm_join_bycondition),
    aes(colour = "Ssd1_target"), 
    size=0.45) +
  geom_point(
    data = left_join(select_genetype_df,
                     tpm_join_bycondition),
    aes(label=Name,colour=GType),
    size=0.45) + 
  ggrepel::geom_text_repel(
    data = left_join(select_genetype_df,
                     tpm_join_bycondition),
    aes(label=Name,colour=GType),
    size=2.5, segment.size=0.1, force=0.15, min.segment.length=0.05,
    # set random number seed so the labels go in the same place every time
    seed = 101) + 
  scale_colour_gtype + theme(legend.position = "none") + 
  scale_x_log10_sci("RNA-seq density (TPM)",
                limits=c(5,max(tpm_join_bycondition$TPM.RNA)),
                oob=scales::squish,expand=c(0.01,0.01)) +
  scale_y_log10_sci("Ssd1 CRAC density (TPM)",
                limits=c(5,max(tpm_join_bycondition$TPM.Ssd1)),
                oob=scales::squish,expand=c(0.01,0.01)) +
  facet_wrap(~Condition,scales="free_y")
plot_Ssd1_vs_RNA_colourlabels

ggsave("../figure_out/Ssd1CRAC_vs_RNAseq_TPM_colourlabels.pdf",
       plot_Ssd1_vs_RNA_colourlabels, width = 6, height = 3.2)
```

# Which proportion of Ssd1-associated reads are in specific Gene categories

### Ssd1-enriched targets have almost half the Ssd1 reads at 30°C, from 4.5% of RNA reads.

```{r Ssd1_targets_totalreadcounts}
Ssd1_targets_DESeq2_2x_p0p05$Gene %>% unique %>% length

tpms_RNA_stranded %>%
  filter(Gene %in% Ssd1_targets_DESeq2_2x_p0p05$Gene) %>%
  summarise_if(is.numeric,sum)
tpms_Ssd1_stranded %>%
  filter(Gene %in% Ssd1_targets_DESeq2_2x_p0p05$Gene) %>%
  summarise_if(is.numeric,sum)

filter(Ssd1_targets_DESeq2_2x_p0p05,Condition == "30°C") %>% nrow()

tpms_RNA_stranded %>%
  filter(Gene %in% filter(Ssd1_targets_DESeq2_2x_p0p05,Condition == "30°C")$Gene) %>%
  summarise_if(is.numeric,sum)
tpms_Ssd1_stranded %>%
  filter(Gene %in% filter(Ssd1_targets_DESeq2_2x_p0p05,Condition == "30°C")$Gene) %>%
  summarise_if(is.numeric,sum)
```

This says that, of the 189 genes which are classified as enriched in Ssd1 binding at 30°C, they have 4.5% of the RNA reads and 43% (mean) of Ssd1-associated reads at 30°C. 

The proportion is smaller at 42°C, and their number of read counts go down.


## About 35% of Ssd1 reads are on Cell Wall proteins, by GO term

```{r load_cell_wall_enrichment}
GOterms_process <- read_tsv("../results/GOterms/Ssd1_targets_DeSeq2_30C_process_GOenriched.txt")
GOterms_component <- read_tsv("../results/GOterms/Ssd1_targets_DeSeq2_30C_component_GOenriched.txt")

cworg_genes <- GOterms_process %>%
  filter(TERM == "cell wall organization") %>%
  pull(ANNOTATED_GENES) %>%
  str_split(", ") %>%
  unlist()

cperi_genes <- GOterms_component %>%
  filter(TERM == "cell periphery") %>%
  pull(ANNOTATED_GENES) %>%
  str_split(", ") %>%
  unlist() %>%
  unique()

tpms_RNA_stranded %>%
  filter(Gene %in% cworg_genes) %>%
  summarise_if(is.numeric,sum)

tpms_Ssd1_stranded %>%
  filter(Gene %in% cworg_genes) %>%
  summarise_if(is.numeric,sum)

tpms_RNA_stranded %>%
  filter(Gene %in% cperi_genes) %>%
  summarise_if(is.numeric,sum)

tpms_Ssd1_stranded %>%
  filter(Gene %in% cperi_genes) %>%
  summarise_if(is.numeric,sum)

```

So at 30°C, about 35% of all Ssd1 reads are on these 35 Cell wall genes, which have only 3% of the mRNAs.
At 42°C, it's more like 24% of Ssd1 reads from 2% of mRNAs.

### Expanding the list for other enriched categories doesn't make a huge difference

```{r cellwalletc_enrichment}
cellwallsurfaceperimem_genes <- 
  GOterms_component %>%
  filter(TERM %in% c("fungal-type cell wall", 
                     "cell wall",
                     "cell periphery",
                     "external encapsulating structure",
                     "extracellular region",
                     "cell surface",
                     "anchored component of membrane")
         ) %>%
  pull(ANNOTATED_GENES) %>%
  str_split(", ") %>%
  unlist()

tpms_RNA_stranded %>%
  filter(Gene %in% cellwallsurfaceperimem_genes) %>%
  summarise_if(is.numeric,sum)

tpms_Ssd1_stranded %>%
  filter(Gene %in% cellwallsurfaceperimem_genes) %>%
  summarise_if(is.numeric,sum)
```

## Although vacuole GO terms are enriched, all the high-count genes are also annotated to the cell wall.

```{r vacuole_enrichment}
vacuole_genes <- 
  GOterms_component %>%
  filter(str_detect(TERM,"vacuole")
         ) %>%
  pull(ANNOTATED_GENES) %>%
  str_split(", ") %>%
  unlist()

tpms_Ssd1_stranded %>%
  filter(Gene %in% vacuole_genes) %>%
  summarise_if(is.numeric,sum)

tpms_Ssd1_stranded %>%
  filter(Gene %in% setdiff(vacuole_genes,cworg_genes)) %>%
  summarise_if(is.numeric,sum)
```


## Session Info

Package versions, etc., that ran this file.

```{r session_info,results="show",message=TRUE}
sessionInfo()
```
